File: 210-course-schedule-ii/course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, vector<int>> graph;
        vector<int> inDegree(numCourses, 0);

        for (auto pre : prerequisites) {
            graph[pre[1]].push_back(pre[0]);
            inDegree[pre[0]]++;
        }
        queue<int> q;
        for(int i = 0; i < numCourses; i++) {
            if (inDegree[i] == 0) q.push(i);
        }
        vector<int> res;
        while (!q.empty()) {
            int curr = q.front();
            q.pop();

            res.push_back(curr);

            for (auto adj : graph[curr]) {
                if (--inDegree[adj] == 0) {
                    q.push(adj);
                }
            }
        }
        if (res.size() == numCourses) return res;
        return {};
    }
};

// 1, 0

// you must take 0 before 1;

//         0
//     1       2
//         3



// 1 -> 0;
// 2 -> 0;
// 3 -> 1, 2;

// indegree[1]++;
// indegree[2]++;
// indegree[3]+= 2;

// graph -> 
// 0 -> 1, 2;
// 1 -> 3;
// 2 -> 3;

// start from root which have the most dependecies, and then at the end process leaf

File: 210-course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];
        HashMap<Integer, ArrayList<Integer>> graph = new HashMap<>();

        for (int[] prerequisite : prerequisites) {
            int first = prerequisite[0];
            int sec = prerequisite[1];
            indegree[first]++;
            graph.computeIfAbsent(sec, k -> new ArrayList<>()).add(first);
        }

        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < indegree.length; i++) {
            if (indegree[i] == 0) q.offer(i);
        }

        ArrayList<Integer> res = new ArrayList<>();

        while (!q.isEmpty()) {
            int curr = q.poll();
            res.add(curr);

            for (int next : graph.getOrDefault(curr, new ArrayList<>())) {
                indegree[next]--;
                if (indegree[next] == 0) {
                    q.offer(next);
                }
            }
        }

        int[] intArray = new int[res.size()];

        // 3. Loop and copy (unboxing occurs automatically)
        for (int i = 0; i < res.size(); i++) {
            // Automatic unboxing converts Integer to int
            intArray[i] = res.get(i); 
        }
        if (res.size() != numCourses) {
            return new int[0];
        }   
        return intArray;

    }
}